Feature/added actions async loading and error

zrnkrystnlee opened the pull request:
- created new actions for loading and error handling
- created freezed generic async file
- modified Overview page
- updated the AppState- added Wait
- updated hard-coded strings to constant
- removed uneccessary constants
- tested flutter analyze and dart format
- created new error handling message- shows 'No Pokemon Available'
 - a Snackbar will show an error message                             
 - updated the String and Color constants files"  

new results of loading screen and error message
https://user-images.githubusercontent.com/72370369/227895844-ad68d629-67c5-40ab-a717-64b85d398155.mp4


results of flutter analyze and dart format
![image](https://user-images.githubusercontent.com/72370369/227856235-4bef3c0a-79d3-4b19-a3e1-8e97ba5d9772.png)

zrnkrystnlee committed:
feat: implemented loading and error message
- created new actions for loading and error handling
- created freezed generic async file
- modified Overview page
- updated the AppState- added Wait
- updated hard-coded strings to constant
- removed uneccessary constants
- tested flutter analyze and dart format

zrnkrystnlee committed:
feat: implemented loading and error message
- created new error handling message- shows 'No Pokemon Available'
- a Snackbar will show an error message
- updated the String and Color constants files

zrnkrystnlee committed:
feat: implemented loading and error message
(pr feedback)
- changed GlobalKey to ScaffoldMessenger.of(context).showSnackBar(SnackBar)

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,28 +1,68 @@
+        ),
+        error: (errorMessage) {
+          WidgetsBinding.instance.addPostFrameCallback((_) {
+            ScaffoldMessenger.of(context).showSnackBar(
```
kindly place the whole of this in a separate private function then name it as `_showErrorMessageSnackbar` 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,28 +1,68 @@
+          WidgetsBinding.instance.addPostFrameCallback((_) {
+            ScaffoldMessenger.of(context).showSnackBar(
+              SnackBar(
+                content: Text(errorMessage!),
```
use null coalescing here instead of enforcing it, to make it safer 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,28 +1,68 @@
+            ScaffoldMessenger.of(context).showSnackBar(
+              SnackBar(
+                content: Text(errorMessage!),
+                duration: const Duration(days: 1),
```
default value for this is already enough no need to customize 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,28 +1,68 @@
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: [
+                const Text(noPokemonsAvailableLabel),
+                const SizedBox(height: 20),
```
maybe you can create a reusable widgets for height and width spacings instead of doing manually every time 🙂

name those widgets as:
- VerticalSpace
- HorizontalSpace

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_vm.dart:
```diff
@@ -1,19 +1,27 @@
 class PokemonOverviewVmFactory extends VmFactory<AppState, PokemonOverviewConnector> {
   @override
-  Vm? fromStore() => PokemonOverviewVm(pokemons: _pokemons);
+  Vm? fromStore() => PokemonOverviewVm(pokemons: _pokemons());
```
why Vm is nullable? 😝

shawn-ffufm commented on pokedex_async_redux/lib/state/action/actions.dart:
```diff
@@ -0,0 +1,15 @@
+import 'package:async_redux/async_redux.dart';
+import 'package:pokedex_asyn_redux/state/app_state.dart';
+
+///AppState Loading Action
```
add a space and also, comment doesn't give me a clear message on what this action is for 😂

shawn-ffufm commented on pokedex_async_redux/lib/utils/constants.dart:
```diff
@@ -8,10 +8,15 @@ const String pokemonOverviewTitle = 'Pokedex';
+const String pokedexErrorImage = 'assets/images/pokedex-glitch.gif';
+const String errorMessage = 'Uh oh! Pokedex not working; Return to Professor Oak';
+const String noPokemonsAvailableLabel = 'No Pokemons Available!';
+const String closeButton = 'Close';
```
do u still need this one? 🙂

shawn-ffufm reviewed (changes requested):
see pr feedback 🙂

zrnkrystnlee committed:
feat: implemented loading and error message
(pr feedback)
- created a new private function for showing error message in snackbar
- remove default value in snackbar
- created a reusable spacing widgets which has Vertical and Horizontal Spaces
- removed null in VM
- modified documentation in actions.dart
- removed unused constant values

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,30 +1,74 @@
+          ),
+        ),
+          error: (errorMessage) {
+            WidgetsBinding.instance.addPostFrameCallback((_) {
```
u can use fat arrow notation here instead 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,30 +1,74 @@
+        ),
+          error: (errorMessage) {
+            WidgetsBinding.instance.addPostFrameCallback((_) {
+              _showErrorMessageSnackbar(context, errorMessage!);
```
use null coalescing and use emptyString constant as its fallback value instead of using non-null enforce here to be safer 🙂

shawn-ffufm commented on pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -1,30 +1,74 @@
+void _showErrorMessageSnackbar(BuildContext context, String message) {
+  ScaffoldMessenger.of(context).showSnackBar(
+    const SnackBar(
+        content: Text(errorMessage ?? ''),
```
no need for extra trailing comma 🙂

shawn-ffufm reviewed (changes requested):
see minor pr feedback 🙂

zrnkrystnlee committed:
feat: implemented loading and error message
>> (pr feedback)
>> - used fat arrow notation
>> - used null coalescing in errorMessage
>> - added an emptyString constant
>> - removed extra trailing comma

shawn-ffufm reviewed (approved):
lgtm
